File: sort/binary_search.c
/*given sorted array with repetitive element find the first occurrence of key */
#include<stdio.h>
#include<stdlib.h>


// this only does searching

int BinarySearch(int array[], int key, int lb , int ub){

	if(ub <lb){
		return -1;
	}

	int midpoint = lb+(ub-lb)/2;
	if(array[midpoint] == key){
		return midpoint;
	}

	if(key > array[midpoint]){
		return BinarySearch(array, key, midpoint+1, ub);
	}
	else{
		// ub = midpoint-1;
		return BinarySearch(array, key, lb, midpoint-1);
	}
}


// Simple binary search algorithm
int binarySearch(int arr[], int l, int r, int x)
{
    if (r>=l)
    {
        int mid = l + (r - l)/2;
        if (arr[mid] == x)
            return mid;
        if (arr[mid] > x)
            return binarySearch(arr, l, mid-1, x);
        return binarySearch(arr, mid+1, r, x);
    }
    return -1;
}



int main(){
	int array[] = {1,2,3,4,4,4,5,6,7,7,8,8,9,10};
	int key = 7;
	int n = sizeof(array)/sizeof(array[0]);
	int result =BinarySearch(array, key, 0, n);
	if(result==-1)
		printf("%d not found on any location\n", key);
	else
		printf("%d found on %dth location\n", key, result);

	int ans =binarySearch(array, 0, n, key);
	if(ans==-1)
		printf("%d not found on any location\n", key);
	else
		printf("%d found on %dth location\n", key, ans);


}

File: Coursera/Algorithms_I/binary_search.py
"""
Find index of a key if it exists in a sorted array

Binary search recurrence

T(N) <= T(N/2) + 1 for N > 1 with T(1) = 1


"""

def binary_search_iter(nums, k):
	l = 0
	r = len(nums)-1
	while l <= r:
		mid = (l+r) / 2
		if nums[mid] > k:
			r = mid - 1
		elif nums[mid] < k:
			l = mid + 1
		else:
			return mid
	return -1


def binary_search(nums, k, l, r):
	if l <= r:
		mid = (l + r) / 2
		if nums[mid] == k:
			return mid
		if nums[mid] > k:
			return binary_search(nums, k, l, mid-1)
		else:
			return binary_search(nums, k, mid+1, r)
	return -1



print binary_search([1,2,3,4,5,6,7], 3, 0, 6)
print binary_search([1,2,3,4,5,6,7], 8, 0, 6)
print binary_search([6, 13, 14, 25, 33, 43, 51, 53, 64, 72, 84, 93, 95, 96, 97], 33, 0, 14)

print binary_search_iter([6, 13, 14, 25, 33, 43, 51, 53, 64, 72, 84, 93, 95, 96, 97], 97)
print binary_search([6, 13, 14, 25, 33, 43, 51, 53, 64, 72, 84, 93, 95, 96], 96, 0, 13)
